hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +19,331 @@ author: Marton Ele
+
+The type of the Container can be defined with the implementation type (eg. Ratis, EC, Closed) and with additional parameters related to type (eg. replication type of Ratis, or EC algorithm for EC containers).
+
+Today's implementation of Ozone can be described with two storage-classes:
+
+![Current replication with Storage class](storage-class-today.png)
+
+The definition of STANDARD Storage class:
+
+ * *First container type/parameters*: Ratis/THREE replicated containers
+ * *Transitions*: In case of any error or if the container is full, convert to closed containers
+ * *Second container type/parameters*: Closed/THREE container
+
+The definition of REDUCED Storage class:
+
+ * *First container type/parameters*: Ratis/ONE replicated containers
+ * *Transitions*: In case the container is full, convert to closed containers
+ * *Second container type/parameters*: Closed/ONE container
+
+This proposal suggests to introduce an abstraction and name the two possible scheme. Later we can define other storage classes as well. For example we can define (Ratis/THREE --> Closed/FIVE) storage class, or more specific containers can be used for Erasure Coding or Random Read/Write.
+
+With this approach the storage-class can be an adjustable abstraction to define the rules of replications. Some key properties of this approach:
+
+**Key properties of the storage-class abstraction**:
+
+ * **Storage-class can be defined by configuration**: Storage class is nothing more just the definition of rules to store / replicate containers. They can be configured in config files and changed any time.
+ * **Object creation requires storage class**: Right now we should defined *replication factor* and *replication type* during the key creation. They can be replaced with setting only the Storage class
+ * **Storage-class defined on key** (and container) **level**
+ * **Storage-class is property of the containers**: As the unit of replication in Ozone is container, one specific storage-class should be adjusted for each containers.
+ * **Changing definition/configuration of storage class** is allowed for most of the properties. The change will modify the behavior of the Replication Manager, and -- eventually -- the containers will be replaced in a different way.
+ * **Changing storage class of a key** is not possible wihout moving data between containers
+ * **Buckets and volumes can provide default storage-classes**, to be inherited if not set on the Key level
+
+*Note*: we already support storage class for S3 objects the only difference is that it would become an Ozone level abstraction and it would defined *all* the container types and transitions.
+
+# Storage-class inside Ozone
+
+First of all, we can configure different replication levels easily with this approach (eg. Ratis/THREE --> Closed/TWO). Ratis need quorum but we can have different replication number after closing containers.
```

Comment:
I don't think that is so easy. A lot of the failure and recovery logic for the Ratis write pipeline assumes that we have three replicas.

Reply:
Agree, it's not easy. But might be possible in case of it's required by users. For the clarity I removed this paragraph.